src/cacao.ts

```diff
@@ -0,0 +1,104 @@
+import * as dagCbor from '@ipld/dag-cbor'
+import * as multiformats from 'multiformats'
+import * as Block from 'multiformats/block'
+import { sha256 as hasher } from 'multiformats/hashes/sha2'
+import { SiweMessage } from './siwe'
+
+export type Header = {
+  t: string
+}
+
+export type Payload = {
+  domain: string
+  iss: string
+  chainId: number
+  aud: string
+  version: string
+  nonce: string
+  iat: number
+  nbf?: number
+  exp?: number
+  statement?: string
+  requestId?: string
+  resources?: Array<string>
+}
+
+export type Signature = {
+  s: string
+}
+
+export type Cacao = {
+  h: Header
+  p: Payload
+  s: Signature
+}
+
+export interface CacaoBlock {
+  value: Cacao
+  cid: multiformats.CID
+  bytes: Uint8Array
+}
+
+class CACAO {
```

Comment:
Suggestion. Instead of going with a CACAO class as a bucket of various things, we can do:

```
export type Cacao = {
  // ... fields
}
export namespace Cacao {
  export function fromSiweMessage(site: string | SiweMessage) {
    // ...
  }
}

export type CacaoBlock = {
  // definition
}

export namespace CacaoBlock {
  export function fromCacao(cacao: Cacao): Promise<CacaoBlock> {
    // ... conversion
  }
}
```

Reply:
hmm, never used namespaces in typescript. how would you differentiate the type and the namespace? or is this kind of like Golang where functions are defined on structs themselves?

Reply:
In TS it's clear when a thing is used as a type or as a value. There are well-defined places where you could only place type, per language semantics. Other places are for values.

Technically, TS types do not exist in JS. `namespace` is compiled though into an object with properties defined inside the namespace.

TS:
```
type Foo = {
  v: string
}

namespace Foo {
	export function blah(input: string) {
      return input
    }
}
``` 

JS:
```
var Foo;
(function (Foo) {
    function blah(input) {
        return input;
    }
    Foo.blah = blah;
})(Foo || (Foo = {}));
```

Reply:
So, it's kind of readability feature given to us by TS.

Reply:
Using namespaces seems to collide with the ESLint configuration for 3box/typescript 
```
  86:8  error  ES2015 module syntax is preferred over custom TypeScript modules and namespaces  @typescript-eslint/no-namespace
  ```

should i just disable that check?

Reply:
> should i just disable that check?

yes, please.